packages/block-editor/src/components/list-view/block-rename-ui.js

```diff
@@ -0,0 +1,163 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	__experimentalInputControl as InputControl,
+	Popover,
+	VisuallyHidden,
+	Button,
+} from '@wordpress/components';
+import { speak } from '@wordpress/a11y';
+import { useInstanceId, useFocusOnMount } from '@wordpress/compose';
+import { useState, forwardRef, useEffect } from '@wordpress/element';
+import { ENTER, ESCAPE } from '@wordpress/keycodes';
+import { __, sprintf } from '@wordpress/i18n';
+import { keyboardReturn, close } from '@wordpress/icons';
+
+const ListViewBlockRenameUI = forwardRef(
+	( { blockTitle, onSubmit, onCancel }, ref ) => {
+		const inputRef = useFocusOnMount();
+
+		const inputDescriptionId = useInstanceId(
+			ListViewBlockRenameUI,
+			`block-editor-list-view-block-node__input-description`
+		);
+
+		const dialogTitle = useInstanceId(
+			ListViewBlockRenameUI,
+			`block-editor-list-view-block-rename-dialog__title`
+		);
+
+		const dialogDescription = useInstanceId(
+			ListViewBlockRenameUI,
+			`block-editor-list-view-block-rename-dialog__description`
+		);
+
+		// Local state for value of input **pre-submission**.
+		const [ inputValue, setInputValue ] = useState( blockTitle );
+
+		const onKeyDownHandler = ( event ) => {
+			// Trap events to input when editing to avoid
+			// default list view key handing (e.g. arrow
+			// keys for navigation).
+			event.stopPropagation();
+
+			// Handle ENTER and ESC exits editing mode.
+			if ( event.keyCode === ENTER || event.keyCode === ESCAPE ) {
+				if ( event.keyCode === ESCAPE ) {
+					handleCancel();
+				}
+
+				if ( event.keyCode === ENTER ) {
+					handleSubmit();
+				}
+			}
+		};
+
+		const handleCancel = () => {
+			// Reset the input's local state to avoid
+			// stale values.
+			setInputValue( blockTitle );
+
+			onCancel();
+
+			// Must be assertive to immediately announce change.
+			speak( __( 'Leaving block name edit mode' ), 'assertive' );
+		};
+
+		const handleSubmit = () => {
+			let successAnnouncement;
+
+			if ( inputValue === '' ) {
+				successAnnouncement = __( 'Block name reset.' );
+			} else {
+				successAnnouncement = sprintf(
+					/* translators: %s: new name/label for the block */
+					__( 'Block name updated to: "%s".' ),
+					inputValue
+				);
+			}
+
+			// Must be assertive to immediately announce change.
+			speak( successAnnouncement, 'assertive' );
+
+			// Submit changes only for ENTER.
+			onSubmit( inputValue );
+		};
+
+		const autoSelectInputText = ( event ) => event.target.select();
+
+		useEffect( () => {
+			speak( __( 'Entering block name edit mode' ), 'assertive' );
+		}, [] );
+
+		return (
+			<Popover
+				anchorRef={ ref }
+				placement="overlay"
+				resize={ true }
+				variant="unstyled"
+				animate={ false }
+				className="block-editor-list-view-block-rename__popover"
+				role="dialog"
+				aria-labelledby={ dialogTitle }
+				aria-describedby={ dialogDescription }
+				onClose={ handleCancel }
+			>
```

Comment:
Do we want this dialog to have a `modal` or `non-modal` behavior? 
Currently, tabbing is constrained within this dialog but it is possible to exit the dialog by arrow keys.
If we want a modal behavior, this dialog should have an `aria-modal="true'` attribute. However, I'm skeptical in giving this a `dialog` role in the first place, because _it doesn't look like a dialog_.

Reply:
Yes, modal is correct here. I do not want users to get the impression they can navigate away.

Reply:
Would using the `Modal` component instead of `Popover` be a better solution? It should implement all of the `modal`-related functionality

Reply:
That sounds ideal but that component is highly opinionated regarding it's visual styling which makes this "inline" editing approach extremely difficult to achieve.

What we really need is a way to disable all styling and just leave the stacking and perhaps the backdrop in place. Is that doable?

Reply:
I understand perfectly — in fact, we're in the process of [discussing a new potential component that would replace `Modal`](https://github.com/WordPress/gutenberg/discussions/56231)

> What we really need is a way to disable all styling and just leave the stacking and perhaps the backdrop in place. Is that doable?

It's _doable_, but very hacky and prone to breaking as soon as `Modal` updates any of its internal implementation.

For that much, it would be probably better "steal" the relevant parts from `Modal` and add them to this particular implementation.

> Aside: if we decoupled the behaviour of all our components into hooks and allow implementors to provide "attach" those behaviours to custom UI ("bring your own markup") then it would be easier to achieve this kind of thing. Is that something we've explored?

This was a behavior that was explored on a few components ([example](https://github.com/WordPress/gutenberg/blob/4b10f653ba12baebd374dacdb2df2d93c017a422/packages/components/src/v-stack/hook.ts)), but ultimately we saw little to no consumers taking advantage of it, and instead we received feedback about the approach just introducing one more layer of unnecessary complexity to our components.